File: 04-hash-tables/02-collision-resolution/c/chaining.c
/**
 * 鏈結法雜湊表（Chaining Hash Table）- C 語言實作
 * 使用鏈結串列處理碰撞，包含鏈長統計
 *
 * Chaining Hash Table implementation with chain length statistics
 */

#include "chaining.h"
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

// ========== 私有函式 Private Functions ==========

/**
 * 計算雜湊值（使用 djb2 演算法）
 * Compute hash value using djb2 algorithm
 *
 * @param key 要雜湊的字串
 * @param capacity 桶的數量
 * @return 桶的索引
 */
static size_t hash(const char* key, size_t capacity) {
    unsigned long hash = 5381;
    int c;

    while ((c = *key++)) {
        // hash * 33 + c
        hash = ((hash << 5) + hash) + c;
    }

    return hash % capacity;
}

/**
 * 建立新的節點
 * Create a new entry
 */
static ChainingEntry* create_entry(const char* key, void* value) {
    ChainingEntry* entry = malloc(sizeof(ChainingEntry));
    if (!entry) return NULL;

    // 複製 key 字串 - Copy key string
    entry->key = strdup(key);
    if (!entry->key) {
        free(entry);
        return NULL;
    }

    entry->value = value;
    entry->next = NULL;
    return entry;
}

/**
 * 釋放節點
 * Free an entry
 */
static void free_entry(ChainingEntry* entry) {
    if (entry) {
        free(entry->key);
        // 注意：不釋放 value，由呼叫者負責
        // Note: value is not freed, caller's responsibility
        free(entry);
    }
}

/**
 * 計算單一鏈結串列的長度
 * Calculate length of a single chain
 */
static size_t chain_length(ChainingEntry* entry) {
    size_t length = 0;
    while (entry) {
        length++;
        entry = entry->next;
    }
    return length;
}

/**
 * 擴容並重新雜湊
 * Resize and rehash
 */
static bool rehash(ChainingHashTable* ht) {
    size_t old_capacity = ht->capacity;
    ChainingEntry** old_buckets = ht->buckets;

    // 容量加倍 - Double capacity
    size_t new_capacity = old_capacity * 2;
    ChainingEntry** new_buckets = calloc(new_capacity, sizeof(ChainingEntry*));
    if (!new_buckets) return false;

    // 更新雜湊表 - Update hash table
    ht->buckets = new_buckets;
    ht->capacity = new_capacity;
    ht->size = 0;

    // 重新插入所有元素 - Reinsert all elements
    for (size_t i = 0; i < old_capacity; i++) {
        ChainingEntry* entry = old_buckets[i];
        while (entry) {
            ChainingEntry* next = entry->next;
            cht_insert(ht, entry->key, entry->value);
            free(entry->key);
            free(entry);
            entry = next;
        }
    }

    free(old_buckets);
    return true;
}

// ========== 公開函式 Public Functions ==========

ChainingHashTable* cht_create(size_t capacity) {
    if (capacity == 0) {
        capacity = CHT_DEFAULT_CAPACITY;
    }

    ChainingHashTable* ht = malloc(sizeof(ChainingHashTable));
    if (!ht) return NULL;

    ht->buckets = calloc(capacity, sizeof(ChainingEntry*));
    if (!ht->buckets) {
        free(ht);
        return NULL;
    }

    ht->capacity = capacity;
    ht->size = 0;
    return ht;
}

void cht_free(ChainingHashTable* ht) {
    if (!ht) return;

    // 釋放所有節點 - Free all entries
    for (size_t i = 0; i < ht->capacity; i++) {
        ChainingEntry* entry = ht->buckets[i];
        while (entry) {
            ChainingEntry* next = entry->next;
            free_entry(entry);
            entry = next;
        }
    }

    free(ht->buckets);
    free(ht);
}

bool cht_insert(ChainingHashTable* ht, const char* key, void* value) {
    if (!ht || !key) return false;

    size_t index = hash(key, ht->capacity);
    ChainingEntry* entry = ht->buckets[index];

    // 檢查 key 是否已存在 - Check if key exists
    while (entry) {
        if (strcmp(entry->key, key) == 0) {
            entry->value = value;  // 更新 - Update existing
            return true;
        }
        entry = entry->next;
    }

    // 建立新節點 - Create new entry
    ChainingEntry* new_entry = create_entry(key, value);
    if (!new_entry) return false;

    // 插入到鏈結串列頭部 - Insert at head
    new_entry->next = ht->buckets[index];
    ht->buckets[index] = new_entry;
    ht->size++;

    // 檢查是否需要擴容 - Check if rehashing needed
    if (cht_load_factor(ht) > CHT_MAX_LOAD_FACTOR) {
        rehash(ht);
    }

    return true;
}

void* cht_search(ChainingHashTable* ht, const char* key) {
    if (!ht || !key) return NULL;

    size_t index = hash(key, ht->capacity);
    ChainingEntry* entry = ht->buckets[index];

    while (entry) {
        if (strcmp(entry->key, key) == 0) {
            return entry->value;
        }
        entry = entry->next;
    }

    return NULL;
}

bool cht_delete(ChainingHashTable* ht, const char* key) {
    if (!ht || !key) return false;

    size_t index = hash(key, ht->capacity);
    ChainingEntry* entry = ht->buckets[index];
    ChainingEntry* prev = NULL;

    while (entry) {
        if (strcmp(entry->key, key) == 0) {
            if (prev) {
                prev->next = entry->next;
            } else {
                ht->buckets[index] = entry->next;
            }
            free_entry(entry);
            ht->size--;
            return true;
        }
        prev = entry;
        entry = entry->next;
    }

    return false;
}

bool cht_contains(ChainingHashTable* ht, const char* key) {
    return cht_search(ht, key) != NULL;
}

size_t cht_size(ChainingHashTable* ht) {
    return ht ? ht->size : 0;
}

size_t cht_capacity(ChainingHashTable* ht) {
    return ht ? ht->capacity : 0;
}

bool cht_is_empty(ChainingHashTable* ht) {
    return ht ? ht->size == 0 : true;
}

double cht_load_factor(ChainingHashTable* ht) {
    if (!ht || ht->capacity == 0) return 0.0;
    return (double)ht->size / ht->capacity;
}

void cht_clear(ChainingHashTable* ht) {
    if (!ht) return;

    for (size_t i = 0; i < ht->capacity; i++) {
        ChainingEntry* entry = ht->buckets[i];
        while (entry) {
            ChainingEntry* next = entry->next;
            free_entry(entry);
            entry = next;
        }
        ht->buckets[i] = NULL;
    }
    ht->size = 0;
}

void cht_get_stats(ChainingHashTable* ht, ChainStats* stats) {
    if (!ht || !stats) return;

    stats->min_length = SIZE_MAX;
    stats->max_length = 0;
    stats->avg_length = 0.0;
    stats->empty_buckets = 0;

    size_t total_length = 0;
    size_t non_empty_buckets = 0;

    // 遍歷所有桶 - Iterate through all buckets
    for (size_t i = 0; i < ht->capacity; i++) {
        size_t length = chain_length(ht->buckets[i]);

        if (length == 0) {
            stats->empty_buckets++;
        } else {
            non_empty_buckets++;
            total_length += length;
            if (length < stats->min_length) {
                stats->min_length = length;
            }
            if (length > stats->max_length) {
                stats->max_length = length;
            }
        }
    }

    // 計算平均鏈長 - Calculate average chain length
    if (non_empty_buckets > 0) {
        stats->avg_length = (double)total_length / non_empty_buckets;
    }

    // 如果沒有非空桶，將最小長度設為 0
    // If no non-empty buckets, set min_length to 0
    if (stats->min_length == SIZE_MAX) {
        stats->min_length = 0;
    }
}

void cht_print_stats(ChainingHashTable* ht) {
    if (!ht) {
        printf("ChainingHashTable: NULL\n");
        return;
    }

    ChainStats stats;
    cht_get_stats(ht, &stats);

    printf("鏈結法雜湊表統計 Chaining Hash Table Statistics:\n");
    printf("  元素數量 Size: %zu\n", ht->size);
    printf("  桶數量 Capacity: %zu\n", ht->capacity);
    printf("  負載因子 Load Factor: %.3f\n", cht_load_factor(ht));
    printf("  空桶數量 Empty Buckets: %zu (%.1f%%)\n",
           stats.empty_buckets,
           100.0 * stats.empty_buckets / ht->capacity);
    printf("  最短鏈長 Min Chain: %zu\n", stats.min_length);
    printf("  最長鏈長 Max Chain: %zu\n", stats.max_length);
    printf("  平均鏈長 Avg Chain: %.2f\n", stats.avg_length);
}

void cht_print(ChainingHashTable* ht) {
    if (!ht) {
        printf("ChainingHashTable: NULL\n");
        return;
    }

    printf("ChainingHashTable (size=%zu, capacity=%zu, load=%.3f):\n",
           ht->size, ht->capacity, cht_load_factor(ht));

    for (size_t i = 0; i < ht->capacity; i++) {
        ChainingEntry* entry = ht->buckets[i];
        if (entry) {
            printf("  [%zu]: ", i);
            while (entry) {
                printf("(%s, %p)", entry->key, entry->value);
                if (entry->next) printf(" -> ");
                entry = entry->next;
            }
            printf("\n");
        }
    }
}

File: 04-hash-tables/02-collision-resolution/javascript/chaining.js
/** Doc block start
 * 鏈結法雜湊表（Chained Hash Table）- JavaScript 實作
 * 使用鏈結法（chaining）處理碰撞，並追蹤探測次數
 *(blank line)
 * Chained Hash Table implementation with collision resolution
 * Using chaining method and tracking probe statistics
 */  // End of block comment

class ChainedHashTable {  // Declare the HashTable class that provides insert/search/delete operations.
    // ========== 常數 Constants ==========
    static DEFAULT_CAPACITY = 16;  // Assign or update a variable that represents the current algorithm state.
    static MAX_LOAD_FACTOR = 0.75;  // Assign or update a variable that represents the current algorithm state.

    /** Doc block start
     * 建構子：初始化鏈結雜湊表 / Constructor: Initialize chained hash table
     *(blank line)
     * @param {number} capacity - 桶的數量（number of buckets）
     */  // End of block comment
    constructor(capacity = ChainedHashTable.DEFAULT_CAPACITY) {  // Assign or update a variable that represents the current algorithm state.
        if (capacity <= 0) {  // Evaluate the condition and branch into the appropriate code path.
            throw new Error('容量必須為正整數 / Capacity must be positive');  // Throw an exception to signal an invalid argument or operation.
        }  // Close the current block scope.
        this._capacity = capacity;  // Assign or update a variable that represents the current algorithm state.
        this._size = 0;  // Assign or update a variable that represents the current algorithm state.
        // 使用陣列的陣列來實作鏈結法 / Use array of arrays for chaining
        this._buckets = new Array(capacity).fill(null).map(() => []);  // Allocate the bucket array and initialize empty chains for each bucket.

        // 統計資訊 - Statistics
        this._totalProbes = 0;  // 總探測次數 - Total number of probes
        this._operations = 0;    // 總操作次數 - Total number of operations
    }  // Close the current block scope.

    // ========== 私有方法 Private Methods ==========

    /** Doc block start
     * 計算雜湊值 / Compute hash value for the given key
     *(blank line)
     * @param {*} key - 要雜湊的鍵
     * @returns {number} 桶的索引（bucket index）
     */  // End of block comment
    _hash(key) {  // Compute a hash-based index so keys map into the table's storage.
        // 將 key 轉為字串後計算雜湊值
        // Convert key to string and compute hash
        const strKey = String(key);  // Assign or update a variable that represents the current algorithm state.
        let hash = 0;  // Assign or update a variable that represents the current algorithm state.

        for (let i = 0; i < strKey.length; i++) {  // Iterate over a range/collection to process each item in sequence.
            // 使用常見的雜湊算法（djb2 變體）
            // Using common hash algorithm (djb2 variant)
            hash = ((hash << 5) - hash + strKey.charCodeAt(i)) | 0;  // Compute a hash-based index so keys map into the table's storage.
        }  // Close the current block scope.

        // 確保回傳正數索引 - Ensure positive index
        return Math.abs(hash) % this._capacity;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 擴容並重新雜湊 / Resize and rehash
     */  // End of block comment
    _rehash() {  // Rehash entries into a larger table to keep operations near O(1) on average.
        const oldBuckets = this._buckets;  // Access or update the bucket storage used to hold entries or chains.
        this._capacity *= 2;  // Grow the table (often doubling) to reduce the load factor and collisions.
        this._buckets = new Array(this._capacity).fill(null).map(() => []);  // Allocate the bucket array and initialize empty chains for each bucket.
        this._size = 0;  // Assign or update a variable that represents the current algorithm state.

        // 重新插入所有元素 - Reinsert all elements
        for (const bucket of oldBuckets) {  // Iterate over a range/collection to process each item in sequence.
            for (const [key, value] of bucket) {  // Iterate over a range/collection to process each item in sequence.
                // 直接插入，不計入統計 - Direct insert without stats
                const index = this._hash(key);  // Compute a hash-based index so keys map into the table's storage.
                this._buckets[index].push([key, value]);  // Access or update the bucket storage used to hold entries or chains.
                this._size++;  // Execute this statement as part of the data structure implementation.
            }  // Close the current block scope.
        }  // Close the current block scope.
    }  // Close the current block scope.

    // ========== 公開方法 Public Methods ==========

    /** Doc block start
     * 插入鍵值對（若 key 已存在則更新）
     * Insert key-value pair (update if key exists)
     *(blank line)
     * 時間複雜度 Time Complexity: 平均 O(1), 最差 O(n)
     *(blank line)
     * @param {*} key - 鍵
     * @param {*} value - 值
     * @returns {number} 探測次數 - Number of probes
     */  // End of block comment
    insert(key, value) {  // Execute this statement as part of the data structure implementation.
        const index = this._hash(key);  // Compute a hash-based index so keys map into the table's storage.
        const bucket = this._buckets[index];  // Access or update the bucket storage used to hold entries or chains.
        let probes = 0;  // Advance or track the probing sequence used by open addressing.

        // 檢查 key 是否已存在 - Check if key exists
        for (let i = 0; i < bucket.length; i++) {  // Iterate over a range/collection to process each item in sequence.
            probes++;  // Advance or track the probing sequence used by open addressing.
            if (bucket[i][0] === key) {  // Evaluate the condition and branch into the appropriate code path.
                bucket[i][1] = value; // 更新 - Update existing
                this._totalProbes += probes;  // Advance or track the probing sequence used by open addressing.
                this._operations++;  // Execute this statement as part of the data structure implementation.
                return probes;  // Return the computed result to the caller.
            }  // Close the current block scope.
        }  // Close the current block scope.

        // 新增鍵值對 - Add new key-value pair
        probes++; // 新增位置也算一次探測 - Adding also counts as a probe
        bucket.push([key, value]);  // Access or update the bucket storage used to hold entries or chains.
        this._size++;  // Execute this statement as part of the data structure implementation.
        this._totalProbes += probes;  // Advance or track the probing sequence used by open addressing.
        this._operations++;  // Execute this statement as part of the data structure implementation.

        // 檢查是否需要擴容 - Check if rehashing needed
        if (this.loadFactor > ChainedHashTable.MAX_LOAD_FACTOR) {  // Evaluate the condition and branch into the appropriate code path.
            this._rehash();  // Rehash entries into a larger table to keep operations near O(1) on average.
        }  // Close the current block scope.

        return probes;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 搜尋給定 key 的 value
     * Search for value associated with key
     *(blank line)
     * 時間複雜度 Time Complexity: 平均 O(1), 最差 O(n)
     *(blank line)
     * @param {*} key - 要搜尋的鍵
     * @returns {{value: *, probes: number}|null} 找到則回傳物件，否則回傳 null
     */  // End of block comment
    search(key) {  // Execute this statement as part of the data structure implementation.
        const index = this._hash(key);  // Compute a hash-based index so keys map into the table's storage.
        const bucket = this._buckets[index];  // Access or update the bucket storage used to hold entries or chains.
        let probes = 0;  // Advance or track the probing sequence used by open addressing.

        for (const [k, v] of bucket) {  // Iterate over a range/collection to process each item in sequence.
            probes++;  // Advance or track the probing sequence used by open addressing.
            if (k === key) {  // Evaluate the condition and branch into the appropriate code path.
                this._totalProbes += probes;  // Advance or track the probing sequence used by open addressing.
                this._operations++;  // Execute this statement as part of the data structure implementation.
                return { value: v, probes };  // Return the computed result to the caller.
            }  // Close the current block scope.
        }  // Close the current block scope.

        // 未找到 - Not found
        this._totalProbes += probes;  // Advance or track the probing sequence used by open addressing.
        this._operations++;  // Execute this statement as part of the data structure implementation.
        return null;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 刪除指定的鍵值對 / Delete key-value pair
     *(blank line)
     * 時間複雜度 Time Complexity: 平均 O(1), 最差 O(n)
     *(blank line)
     * @param {*} key - 要刪除的鍵
     * @returns {{success: boolean, probes: number}} 刪除結果與探測次數
     */  // End of block comment
    delete(key) {  // Execute this statement as part of the data structure implementation.
        const index = this._hash(key);  // Compute a hash-based index so keys map into the table's storage.
        const bucket = this._buckets[index];  // Access or update the bucket storage used to hold entries or chains.
        let probes = 0;  // Advance or track the probing sequence used by open addressing.

        for (let i = 0; i < bucket.length; i++) {  // Iterate over a range/collection to process each item in sequence.
            probes++;  // Advance or track the probing sequence used by open addressing.
            if (bucket[i][0] === key) {  // Evaluate the condition and branch into the appropriate code path.
                bucket.splice(i, 1);  // Access or update the bucket storage used to hold entries or chains.
                this._size--;  // Execute this statement as part of the data structure implementation.
                this._totalProbes += probes;  // Advance or track the probing sequence used by open addressing.
                this._operations++;  // Execute this statement as part of the data structure implementation.
                return { success: true, probes };  // Return the computed result to the caller.
            }  // Close the current block scope.
        }  // Close the current block scope.

        // 未找到 - Not found
        this._totalProbes += probes;  // Advance or track the probing sequence used by open addressing.
        this._operations++;  // Execute this statement as part of the data structure implementation.
        return { success: false, probes };  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 檢查 key 是否存在
     * Check if key exists
     *(blank line)
     * @param {*} key - 要檢查的鍵
     * @returns {boolean}
     */  // End of block comment
    contains(key) {  // Execute this statement as part of the data structure implementation.
        return this.search(key) !== null;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 計算負載因子 / Calculate load factor
     *(blank line)
     * @returns {number} 負載因子（元素數 / 桶數）
     */  // End of block comment
    get loadFactor() {  // Execute this statement as part of the data structure implementation.
        return this._size / this._capacity;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 回傳元素數量 / Return number of elements
     *(blank line)
     * @returns {number}
     */  // End of block comment
    get size() {  // Execute this statement as part of the data structure implementation.
        return this._size;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 回傳桶的數量 / Return number of buckets
     *(blank line)
     * @returns {number}
     */  // End of block comment
    get capacity() {  // Execute this statement as part of the data structure implementation.
        return this._capacity;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 檢查是否為空 / Check if empty
     *(blank line)
     * @returns {boolean}
     */  // End of block comment
    isEmpty() {  // Execute this statement as part of the data structure implementation.
        return this._size === 0;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 清空雜湊表 / Clear all elements
     */  // End of block comment
    clear() {  // Execute this statement as part of the data structure implementation.
        this._buckets = new Array(this._capacity).fill(null).map(() => []);  // Allocate the bucket array and initialize empty chains for each bucket.
        this._size = 0;  // Assign or update a variable that represents the current algorithm state.
        this._totalProbes = 0;  // Assign or update a variable that represents the current algorithm state.
        this._operations = 0;  // Assign or update a variable that represents the current algorithm state.
    }  // Close the current block scope.

    /** Doc block start
     * 取得平均探測次數 / Get average number of probes per operation
     *(blank line)
     * @returns {number} 平均探測次數
     */  // End of block comment
    getAverageProbes() {  // Execute this statement as part of the data structure implementation.
        return this._operations === 0 ? 0 : this._totalProbes / this._operations;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 取得最長鏈結長度 / Get maximum chain length
     *(blank line)
     * @returns {number} 最長鏈結的長度
     */  // End of block comment
    getMaxChainLength() {  // Execute this statement as part of the data structure implementation.
        let maxLength = 0;  // Assign or update a variable that represents the current algorithm state.
        for (const bucket of this._buckets) {  // Iterate over a range/collection to process each item in sequence.
            if (bucket.length > maxLength) {  // Evaluate the condition and branch into the appropriate code path.
                maxLength = bucket.length;  // Access or update the bucket storage used to hold entries or chains.
            }  // Close the current block scope.
        }  // Close the current block scope.
        return maxLength;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 取得統計資訊 / Get statistics
     *(blank line)
     * @returns {Object} 統計資訊物件
     */  // End of block comment
    getStatistics() {  // Execute this statement as part of the data structure implementation.
        return {  // Return the computed result to the caller.
            size: this._size,  // Execute this statement as part of the data structure implementation.
            capacity: this._capacity,  // Execute this statement as part of the data structure implementation.
            loadFactor: this.loadFactor,  // Execute this statement as part of the data structure implementation.
            totalProbes: this._totalProbes,  // Execute this statement as part of the data structure implementation.
            operations: this._operations,  // Execute this statement as part of the data structure implementation.
            averageProbes: this.getAverageProbes(),  // Execute this statement as part of the data structure implementation.
            maxChainLength: this.getMaxChainLength()  // Execute this statement as part of the data structure implementation.
        };  // Execute this statement as part of the data structure implementation.
    }  // Close the current block scope.

    // ========== 迭代器 Iterators ==========

    /** Doc block start
     * 回傳所有鍵的陣列 / Return array of all keys
     *(blank line)
     * @returns {Array}
     */  // End of block comment
    keys() {  // Execute this statement as part of the data structure implementation.
        const result = [];  // Assign or update a variable that represents the current algorithm state.
        for (const bucket of this._buckets) {  // Iterate over a range/collection to process each item in sequence.
            for (const [key] of bucket) {  // Iterate over a range/collection to process each item in sequence.
                result.push(key);  // Execute this statement as part of the data structure implementation.
            }  // Close the current block scope.
        }  // Close the current block scope.
        return result;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 回傳所有值的陣列 / Return array of all values
     *(blank line)
     * @returns {Array}
     */  // End of block comment
    values() {  // Execute this statement as part of the data structure implementation.
        const result = [];  // Assign or update a variable that represents the current algorithm state.
        for (const bucket of this._buckets) {  // Iterate over a range/collection to process each item in sequence.
            for (const [, value] of bucket) {  // Iterate over a range/collection to process each item in sequence.
                result.push(value);  // Execute this statement as part of the data structure implementation.
            }  // Close the current block scope.
        }  // Close the current block scope.
        return result;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 回傳所有鍵值對的陣列 / Return array of all key-value pairs
     *(blank line)
     * @returns {Array}
     */  // End of block comment
    entries() {  // Execute this statement as part of the data structure implementation.
        const result = [];  // Assign or update a variable that represents the current algorithm state.
        for (const bucket of this._buckets) {  // Iterate over a range/collection to process each item in sequence.
            for (const entry of bucket) {  // Iterate over a range/collection to process each item in sequence.
                result.push([...entry]);  // Execute this statement as part of the data structure implementation.
            }  // Close the current block scope.
        }  // Close the current block scope.
        return result;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 字串表示 / String representation
     *(blank line)
     * @returns {string}
     */  // End of block comment
    toString() {  // Execute this statement as part of the data structure implementation.
        const items = this.entries()  // Assign or update a variable that represents the current algorithm state.
            .map(([k, v]) => `${String(k)}: ${String(v)}`)  // Assign or update a variable that represents the current algorithm state.
            .join(', ');  // Execute this statement as part of the data structure implementation.
        return `ChainedHashTable{${items}}`;  // Return the computed result to the caller.
    }  // Close the current block scope.
}  // Close the current block scope.

// ========== 匯出 Export ==========
// CommonJS 匯出 - CommonJS export
if (typeof module !== 'undefined' && module.exports) {  // Evaluate the condition and branch into the appropriate code path.
    module.exports = ChainedHashTable;  // Assign or update a variable that represents the current algorithm state.
}  // Close the current block scope.

// ========== 範例執行 Example Usage ==========
if (typeof require !== 'undefined' && require.main === module) {  // Evaluate the condition and branch into the appropriate code path.
    console.log('=== 鏈結法雜湊表示範 Chained Hash Table Demo ===\n');  // Execute this statement as part of the data structure implementation.

    // 建立雜湊表 - Create hash table
    const ht = new ChainedHashTable(8);  // Assign or update a variable that represents the current algorithm state.

    // 插入操作 - Insert operations
    console.log('插入鍵值對 Inserting key-value pairs:');  // Execute this statement as part of the data structure implementation.
    const probes1 = ht.insert('apple', 100);  // Advance or track the probing sequence used by open addressing.
    console.log(`  插入 'apple': ${probes1} 次探測`);  // Advance or track the probing sequence used by open addressing.
    const probes2 = ht.insert('banana', 200);  // Advance or track the probing sequence used by open addressing.
    console.log(`  插入 'banana': ${probes2} 次探測`);  // Advance or track the probing sequence used by open addressing.
    const probes3 = ht.insert('cherry', 300);  // Advance or track the probing sequence used by open addressing.
    console.log(`  插入 'cherry': ${probes3} 次探測`);  // Advance or track the probing sequence used by open addressing.
    console.log(`  雜湊表: ${ht}\n`);  // Execute this statement as part of the data structure implementation.

    // 搜尋操作 - Search operations
    console.log('搜尋操作 Search operations:');  // Execute this statement as part of the data structure implementation.
    const result1 = ht.search('apple');  // Assign or update a variable that represents the current algorithm state.
    console.log(`  search('apple') = ${result1.value}, 探測: ${result1.probes}`);  // Advance or track the probing sequence used by open addressing.
    const result2 = ht.search('banana');  // Assign or update a variable that represents the current algorithm state.
    console.log(`  search('banana') = ${result2.value}, 探測: ${result2.probes}`);  // Advance or track the probing sequence used by open addressing.
    const result3 = ht.search('grape');  // Assign or update a variable that represents the current algorithm state.
    console.log(`  search('grape') = ${result3}, 未找到\n`);  // Assign or update a variable that represents the current algorithm state.

    // 刪除操作 - Delete operation
    console.log('刪除操作 Delete operation:');  // Execute this statement as part of the data structure implementation.
    const delResult = ht.delete('banana');  // Assign or update a variable that represents the current algorithm state.
    console.log(`  刪除 'banana': 成功=${delResult.success}, 探測=${delResult.probes}\n`);  // Advance or track the probing sequence used by open addressing.

    // 統計資訊 - Statistics
    console.log('統計資訊 Statistics:');  // Execute this statement as part of the data structure implementation.
    const stats = ht.getStatistics();  // Assign or update a variable that represents the current algorithm state.
    console.log(`  大小 Size: ${stats.size}`);  // Execute this statement as part of the data structure implementation.
    console.log(`  容量 Capacity: ${stats.capacity}`);  // Execute this statement as part of the data structure implementation.
    console.log(`  負載因子 Load Factor: ${stats.loadFactor.toFixed(3)}`);  // Execute this statement as part of the data structure implementation.
    console.log(`  總探測次數 Total Probes: ${stats.totalProbes}`);  // Execute this statement as part of the data structure implementation.
    console.log(`  總操作次數 Operations: ${stats.operations}`);  // Execute this statement as part of the data structure implementation.
    console.log(`  平均探測次數 Average Probes: ${stats.averageProbes.toFixed(3)}`);  // Execute this statement as part of the data structure implementation.
    console.log(`  最長鏈結 Max Chain: ${stats.maxChainLength}`);  // Execute this statement as part of the data structure implementation.
}  // Close the current block scope.
